File: python-ver/tile.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from random import shuffle
from util import repeat

########################
### Data Definitions ###
########################

regular_tiles = (
      ('C', 1), ('C', 2), ('C', 3), ('C', 4), ('C', 5)
    , ('C', 6), ('C', 7), ('C', 8), ('C', 9)
    , ('B', 1), ('B', 2), ('B', 3), ('B', 4), ('B', 5)
    , ('B', 6), ('B', 7), ('B', 8), ('B', 9)
    , ('K', 1), ('K', 2), ('K', 3), ('K', 4), ('K', 5)
    , ('K', 6), ('K', 7), ('K', 8), ('K', 9)
    , ('W', 1), ('W', 2), ('W', 3), ('W', 4)
    , ('D', 1), ('D', 2), ('D', 3)
    )

bonus_tiles = (
      ('F', 1), ('F', 2), ('F', 3), ('F', 4)
    , ('S', 1), ('S', 2), ('S', 3), ('S', 4)
    )

def get_wall():
    w = []
    for t in regular_tiles:
        w += repeat(t, 4)
    for t in bonus_tiles:
        w.append(t)
    return random.shuffle(w)



###################
### Conversions ###
###################

def show_tile(tile):
    return fst(tile) + str( snd(tile) )

def read_tile(tile):
    if len(tile) == 2:
        return (tile[0], int( tile[1] ) )

rank = { 'C': 10
       , 'B': 20
       , 'K': 30
       , 'W': 40
       , 'D': 50
       , 'F': 60
       , 'S': 70
       , 'A': 80
       }

def get_rank(tile):
    return rank[ fst(tile) ] + snd(tile)

def compare(tile1, tile2):
    if get_rank(tile1) < get_rank(tile2):
        return -1
    elif get_rank(tile1) > get_rank(tile2):
        return 1
    return 0



###########################
### Queries about Tiles ###
###########################

def is_coin(tile):
    if fst(tile) == 'C':
        return True
    return False

def is_bamboo(tile):
    if fst(tile) == 'B':
        return True
    return False

def is_character(tile):
    if fst(tile) == 'K':
        return True
    return False

def is_suit(tile):
    return is_coin(tile) or is_bamboo(tile) or is_character(tile)

def is_simple(tile):
    if is_suit(tile):
        s = snd(tile)
        if s > 1 and s < 9:
            return True
    return False

def is_terminal(tile):
    if is_suit(tile):
        s = snd(tile)
        if s == 1 or s == 9:
            return True
    return False

def is_wind(tile):
    if fst(tile) == 'W':
        return True
    return False

def is_dragon(tile):
    if fst(tile) == 'D':
        return True
    return False

def is_honor(tile):
    return is_wind(tile) or is_dragon(tile)

def is_edge(tile):
    return is_terminal(tile) or is_honor(tile)

def is_flower(tile):
    if fst(tile) == 'F':
        return True
    return False

def is_season(tile):
    if fst(tile) == 'S':
        return True
    return False

# could probably check if tile is in bonus_tile too
# but that seems way more comparisons than just looking at the suit
def is_bonus(tile):
    return is_flower(tile) or is_season(tile)

def is_green(tile):
    green = [('B', 2), ('B', 3), ('B', 4), ('B', 6), ('B', 8), ('D', 2)]
    return tile in green

def is_red(tile):
    red = [('B', 1), ('B', 5), ('B', 7), ('B', 9), ('D', 1)]
    return tile in red

def is_blue(tile):
    blue = [('C', 8), ('W', 1), ('W', 2), ('W', 3), ('W', 4), ('D', 3)]
    return tile in blue



###########################
### Utilities functions ###
###########################

def fst(tile):
    return tile[0]

def snd(tile):
    return tile[1]

def succ(tile):
    if is_suit(tile):
        if snd(tile) < 9:
            return (fst(tile), snd(tile) + 1)
    return None

def pred(tile):
    if is_suit(tile):
        if snd(tile) > 1:
            return (fst(tile), snd(tile) - 1)
    return None

File: Game/Mahjong/Tile.hs
-- |
-- Module      :  Game.Mahjong.Tile
-- Copyright   :  Joseph Ching 2014
-- License     :  MIT
--
-- Maintainer  :  joseph.m.ching@gmail.com
-- Stability   :  experimental
-- Portability :  portable

-- | Tile implementation and tile related functions
module Game.Mahjong.Tile (
      -- Tile data
      SuitValue, WindValue, DragonValue, FlowerValue, SeasonValue
    , Tile(..)

      -- Tile collections
    , coins, bamboos, characters, winds, dragons, flowers, seasons
    , suits, simples, terminals, honors, edges, bonuses
    , greens, reds, blues
    , regulars, alls

      -- Tile predicates
    , isCoin, isBamboo, isCharacter, isWind, isDragon, isFlower, isSeason
    , isSuit, isSimple, isTerminal, isHonor, isEdge, isBonus
    , isGreen, isRed, isBlue
    
      -- Utility functions
    , mjSet, getWall, impureWall
    , dora, reverseDora
    ) where

import Data.List (intersperse)
import Data.Map (Map, insert, (!), elems, singleton)
import Data.Maybe (fromJust)
import System.IO.Unsafe
import System.Random


{- Data definition -}

-- | The nine suit values used for the three suits
data SuitValue = One | Two | Three | Four | Five | Six | Seven | Eight | Nine
                 deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- | The four wind values
data WindValue = East | South | West | North
                 deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- | The three dragon values
data DragonValue = Red | Green | White
                   deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- | The four flower values
data FlowerValue = PlumBlossom | Orchid | Chrysanthenum | BambooTree
                   deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- | The four season values
data SeasonValue = Spring | Summer | Autumn | Winter
                   deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- | The four animal values
data AnimalValue = Cat | Mouse | Cockerel | Centipede
                   deriving (Bounded, Enum, Eq, Ord, Read, Show)

-- | Define the tiles
data Tile = Coin SuitValue
          | Bamboo SuitValue
          | Character SuitValue
          | Wind WindValue
          | Dragon DragonValue
          | Flower FlowerValue
          | Season SeasonValue
          | Animal AnimalValue
            deriving (Eq, Ord, Read)

instance Show Tile where
  show = showTile 

showTile :: Tile -> String
showTile t =
  case t of
    Coin c      -> 'C' : (getValue c [One ..])
    Bamboo b    -> 'B' : (getValue b [One ..])
    Character k -> 'K' : (getValue k [One ..])
    Wind w      -> 'W' : (getValue w [East ..])
    Dragon d    -> 'D' : (getValue d [Red ..])
    Flower f    -> 'F' : (getValue f [PlumBlossom ..])
    Season s    -> 'S' : (getValue s [Spring ..])
    Animal a    -> 'A' : (getValue a [Cat ..])
    where
      getValue :: (Eq a) => a -> [a] -> String
      getValue v = show . fromJust . lookup v . flip zip [1..]


{- Tile collections -}

coins :: [Tile]
coins = map (Coin $) [One ..]

bamboos :: [Tile]
bamboos = map (Bamboo $) [One ..] 

characters :: [Tile]
characters = map (Character $) [One ..] 

winds :: [Tile]
winds = map (Wind $) [East ..]

dragons :: [Tile]
dragons = map (Dragon $) [Red ..] 

flowers :: [Tile]
flowers = map (Flower $) [PlumBlossom ..] 

seasons :: [Tile]
seasons = map (Season $) [Spring ..] 

animals :: [Tile]
animals = map (Animal $) [Cat ..]

suits :: [Tile]
suits = coins ++ bamboos ++ characters

simples :: [Tile]
simples = concatMap (\x -> tail . init $ x) [coins, bamboos, characters]

terminals :: [Tile]
terminals = concatMap (\x -> [head x, last x]) [coins, bamboos, characters]

honors :: [Tile]
honors = winds ++ dragons

edges :: [Tile]
edges = terminals ++ honors

bonuses :: [Tile]
bonuses = flowers ++ seasons

extras :: [Tile]
extras = bonuses ++ animals

greens :: [Tile]
greens = map (Bamboo $) [Two, Three, Four, Six, Eight] ++ [Dragon Green]

reds :: [Tile]
reds = map (Bamboo $) [One, Five, Seven, Nine] ++ [Dragon Red]

blues :: [Tile]
blues = [Coin Eight] ++ winds ++ [Dragon White]

regulars :: [Tile]
regulars = coins ++ bamboos ++ characters ++ winds ++ dragons

alls :: [Tile]
alls = regulars ++ bonuses


{- Predicates for determining tile types -}

isCoin :: Tile -> Bool
isCoin (Coin _) = True
isCoin _        = False

isBamboo :: Tile -> Bool
isBamboo (Bamboo _) = True
isBamboo _          = False

isCharacter :: Tile -> Bool
isCharacter (Character _) = True
isCharacter _             = False

isWind :: Tile -> Bool
isWind (Wind _) = True
isWind _        = False

isDragon :: Tile -> Bool
isDragon (Dragon _) = True
isDragon _          = False

isFlower :: Tile -> Bool
isFlower (Flower _) = True
isFlower _          = False

isSeason :: Tile -> Bool
isSeason (Season _) = True
isSeason _          = False

isAnimal :: Tile -> Bool
isAnimal (Animal _) = True
isAnimal _          = False

isSuit :: Tile -> Bool
isSuit = or . zipWith id [isCoin, isBamboo, isCharacter] . repeat

isSimple :: Tile -> Bool
isSimple = and . zipWith id [isSuit, not . isTerminal] . repeat

isTerminal :: Tile -> Bool
isTerminal = flip elem terminals

isHonor :: Tile -> Bool
isHonor = or . zipWith id [isWind, isDragon] . repeat

isEdge :: Tile -> Bool
isEdge = or . zipWith id [isTerminal, isWind, isDragon] . repeat

isBonus :: Tile -> Bool
isBonus = or . zipWith id [isFlower, isSeason, isAnimal] . repeat

isGreen :: Tile -> Bool
isGreen = flip elem greens

isRed :: Tile -> Bool
isRed = flip elem reds

isBlue :: Tile -> Bool
isBlue = flip elem blues


{- Utility functions -}

dora :: Tile -> Tile
dora (Coin c)      = if c == Nine       then Coin One           else Coin $ succ c
dora (Bamboo b)    = if b == Nine       then Bamboo One         else Bamboo $ succ b
dora (Character k) = if k == Nine       then Character One      else Character $ succ k
dora (Wind w)      = if w == North      then Wind East          else Wind $ succ w
dora (Dragon d)    = if d == White      then Dragon Red         else Dragon $ succ d
dora (Flower f)    = if f == BambooTree then Flower PlumBlossom else Flower $ succ f
dora (Season s)    = if s == Winter     then Season Spring      else Season $ succ s
dora (Animal a)    = if a == Centipede  then Animal Cat         else Animal $ succ a

reverseDora :: Tile -> Tile
reverseDora (Coin c)      = if c == One         then Coin Nine         else Coin $ pred c
reverseDora (Bamboo b)    = if b == One         then Bamboo Nine       else Bamboo $ pred b
reverseDora (Character k) = if k == One         then Character Nine    else Character $ pred k
reverseDora (Wind w)      = if w == East        then Wind North        else Wind $ pred w
reverseDora (Dragon d)    = if d == Red         then Dragon White      else Dragon $ pred d
reverseDora (Flower f)    = if f == PlumBlossom then Flower BambooTree else Flower $ pred f
reverseDora (Season s)    = if s == Spring      then Season Winter     else Season $ pred s
reverseDora (Animal a)    = if a == Cat         then Animal Centipede  else Animal $ pred a


{- Wall building -}

mjSet :: [Tile]
mjSet = (concatMap (take 4 . repeat) $ regulars) ++ bonuses

getWall :: Int -> [Tile]
getWall a = fst $ fisherYates (mkStdGen a) mjSet

impureWall :: [Tile]
impureWall = getWall impureRandNumber 

impureRandNumber :: Int
impureRandNumber = mod (unsafePerformIO randomIO) 144

-- | Fisher Yates Algorithm
-- | Source:  http://www.haskell.org/haskellwiki/Random_shuffle
fisherYatesStep :: RandomGen g => (Map Int a, g) -> (Int, a) -> (Map Int a, g)
fisherYatesStep (m, gen) (i, x) = ((insert j x . insert i (m ! j)) m, gen')
  where
    (j, gen') = randomR (0, i) gen

fisherYates :: RandomGen g => g -> [a] -> ([a], g)
fisherYates gen [] = ([], gen)
fisherYates gen l =
  toElems $ foldl fisherYatesStep (initial (head l) gen) (numerate (tail l))
  where
    toElems (x, y) = (elems x, y)
    numerate = zip [1..]
    initial x gen = (singleton 0 x, gen)
